File: csharp/1289.minimum-falling-path-sum-ii.cs
/*
 * @lc app=leetcode id=1289 lang=csharp
 *
 * [1289] Minimum Falling Path Sum II
 */

// @lc code=start
public class Solution {
    public int MinFallingPathSum(int[][] arr) {
        int n = arr.Length;
        (int Value, int Index) min = (Value: 0, Index: -1), min2 = (Value: 0, Index: -1);
        int[] tmp = new int[n];
        foreach (var row in arr) {
            for (int i = 0; i < n; ++i) {
                if (i == min.Index) {
                    tmp[i] = row[i] + min2.Value;
                } else {
                    tmp[i] = row[i] + min.Value;
                }
            }
            min = (Value: 0, Index: -1);
            min2 = (Value: 0, Index: -1);
            for (int i = 0; i < n; ++i) {
                if (min.Index == -1 || min.Value >= tmp[i]) {
                    min2 = min;
                    min = (Value: tmp[i], Index: i);
                } else if (min2.Index == -1 || min2.Value > tmp[i]) {
                    min2 = (Value: tmp[i], Index: i);
                }
            }
        }
        return min.Value;
    }
}
// @lc code=end

File: java/1289.minimum-falling-path-sum-ii.java
/*
 * @lc app=leetcode id=1289 lang=java
 *
 * [1289] Minimum Falling Path Sum II
 */

// @lc code=start
class Solution {
    public int minFallingPathSum(int[][] grid) {
        int n = grid.length, mi1 = -1, mv1 = 0, mv2 = 0;
        for (int i = 0; i < n; ++i) {
            int nmi1 = -1, nmv1 = Integer.MAX_VALUE, nmv2 = Integer.MAX_VALUE;
            for (int j = 0; j < n; ++j) {
                int v = grid[i][j] + (j == mi1 ? mv2 : mv1);
                if (v < nmv1) {
                    nmv2 = nmv1;
                    nmi1 = j;
                    nmv1 = v;
                } else if (v < nmv2) {
                    nmv2 = v;
                }
            }
            mi1 = nmi1;
            mv1 = nmv1;
            mv2 = nmv2;
        }
        return mv1;
    }
}
// @lc code=end

